File: app/schemas/vin.py
from urllib.parse import urlparse
from pydantic import BaseModel, ValidationInfo, field_validator

class Vin(BaseModel):
  """ The domain model that represents a VIN. """
  vin: str
  make: str
  model: str
  model_year: str
  body_class: str
  photo_url: str = ""

  class Config:
    from_attributes = True

  @field_validator('*')
  @classmethod
  def remove_whitespaces(cls, value: str) -> str:
    return value.strip()

  @field_validator('vin')
  @classmethod
  def make_vin_upper_case(cls, value: str) -> str:
    return value.upper()

  @field_validator('photo_url')
  @classmethod
  def is_photo_url_valid(cls, value: str) -> str:
    # Only check if photoUrl is provided
    if len(value) > 0 and not cls.is_url(value):
      raise ValueError(f"photoUrl is not a valid url \"{value}\".")
    return value

  @field_validator('vin')
  @classmethod
  def check_vin_format(cls, value: str) -> str:
    if not cls.is_vin_correct_format(value):
      raise ValueError(f"VIN \"{value}\" must be a 17 alphanumeric characters string.")
    return value

  @field_validator("vin", "make", "model", "model_year", "body_class")
  @classmethod
  def check_field_empty(cls, value: str, info: ValidationInfo) -> str:
    if len(value) == 0:
      raise ValueError(f"Field \"{info.field_name}\" must not be an empty string.")
    return value

  @classmethod
  def is_vin_correct_format(cls, vin: str) -> bool:
    """ Check if the given vin is 17 characters and
        all characters are alphanumeric.
    """
    return len(vin) == 17 and vin.isalnum()

  @classmethod
  def is_url(cls, url: str) -> bool:
    """ Taken from
        https://stackoverflow.com/questions/7160737/how-to-validate-a-url-in-python-malformed-or-not
    """
    try:
      result = urlparse(url)
      return all([result.scheme, result.netloc])
    except ValueError:
      return False

File: nodejs/src/schemas/vin.ts
import { z } from 'zod';

// https://dev.to/isnan__h/custom-schema-validation-in-typescript-with-zod-5cp5
// https://github.com/colinhacks/zod/issues/37

export type Vin = z.infer<typeof VinSchema>;

export function validate(vin: Vin) {
  return VinSchema.parse(vin);
}

export function isVinCorrectFormat(vinNumber: string): boolean {
  return vinNumber.length === 17 && isAlphaNumeric(vinNumber);
}

const VinSchema = z.object({
  vin: z.string({})
    .transform((value, _) => value.trim().toUpperCase())
    .refine(isVinCorrectFormat,
            value => ({ message: `"VIN ${value}" must be a 17 alphanumeric characters string.`})),
  make: z.string()
    .transform((value, _) => value.trim())
    .refine(isNotEmpty, getEmptyStringMessage),
  model: z.string()
    .transform((value, _) => value.trim())
    .refine(isNotEmpty, getEmptyStringMessage),
  modelYear: z.string()
    .transform((value, _) => value.trim())
    .refine(isNotEmpty),
  bodyClass: z.string()
    .transform((value, _) => value.trim())
    .refine(isNotEmpty, getEmptyStringMessage),
  photoUrl: z.string().url().optional(),
});

function isAlphaNumeric(str: string): boolean {
  for (let i = 0; i < str.length; i++) {
    const code = str.charCodeAt(i);
    if (!(code > 47 && code < 58) && // numeric (0-9)
        !(code > 64 && code < 91) && // upper alpha (A-Z)
        !(code > 96 && code < 123)) { // lower alpha (a-z)
      return false;
    }
  }
  return true;
};

function isNotEmpty(str: string) {
  return str !== '' && str !== null && str !== undefined;
}

function getEmptyStringMessage(_: string) {
  return { message: `Value must not be an empty string.`};
}
